picocli-shell-jline3/src/main/java/picocli/tool/jline3/ExampleTool.java

```diff
@@ -0,0 +1,146 @@
+
+        // Use case #1 -- process command line arguments
+        if (!arguments.isEmpty()) {
+            for (final String argument : arguments) {
+                final Integer result = process(argument);
+                if (0 != result) return result;
+            }
+            return 0;
+        }
+        // Use case #2 -- pipe or shell redirection
+        else if (null == System.console()) return runFromInput();
+        // Use case #3 -- interactive REPL
+        else return runRepl(prompt);
+    }
+
+    /** @todo Nicer way using more modern Java. */
+    private static Integer runFromInput() {
+        try (final BufferedReader lineReader = lineReader()) {
+            while (true) {
+                final String line = lineReader.readLine();
+                if (null == line) break;
+                final Integer result = process(line);
+                if (0 != result) return result;
+            }
+        } catch (final IOException e) {
+            return 2;
+        }
+        return 0;
+    }
+
+    private static BufferedReader lineReader() {
+        return new BufferedReader(new InputStreamReader(System.in));
+    }
+}
+
+final class Repl {
+    /**
+     * This is an <em>absolutely minimal</em> REPL.
+     *
+     * @param prompt the prompt to show on left-side asking user for input
+     *
+     * @return the exit code for <code>main()</code>
+     *
+     * @todo Newer versions of Java permit defining this as a lambda
+     */
+    static Integer runRepl(final String prompt) {
+        try (final Terminal terminal = TerminalBuilder.builder().name("example tool").build()) {
+            final LineReader lineReader = LineReaderBuilder.builder().terminal(terminal).build();
+            do {
+                final String input;
+                try {
+                    input = lineReader.readLine(prompt);
+                } catch (final UserInterruptException | EndOfFileException e) {
+                    // Handling user banging on <kbd>Ctrl-C</kbd>
+                    // (<code>UserInterruptException</code>), or ending
+                    // input with <kbd>Ctrl-D</kbd> or <kbd>Ctrl-Z</kbd>
+                    // (<code>EndOfFileException</code>) is a policy
+                    // decision for your program.  However you handle
+                    // these, you need a means to exit the "do-while" loop.
+                    return 0;
+                }
+                // TODO: It is challenging in Java to return _both_ a status
+                //  code for the command line, and to let caller handle
+                //  exceptions
+                final Integer result = process(input);
+                // Another policy decision: Should invalid input kill the
+                // REPL, or notify the user and continue?
+                if (0 != result) return result;
+            } while (true);
+        } catch (final IOException e) {
+            // TODO: It is challenging in Java to return _both_ a status code
```

Comment:
If processing the exception is challenging them perhaps we should not catch it here, or perhaps we should wrap it in a custom exception and re-throw it .